src/Nethermind/Nethermind.Blockchain/HistoryPruning/HistoryPruner.cs

```diff
@@ -0,0 +1,269 @@
+                Hash256 hash = block.Hash;
+
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    if (_logger.IsInfo) _logger.Info($"Pruning operation timed out at timestamp {cutoffTimestamp}. Deleted {deletedBlocks} blocks.");
+                    break;
+                }
+
+                if (number == 0 || number >= _blockTree.SyncPivot.BlockNumber)
+                {
+                    continue;
+                }
+
+                // todo: change to debug after testing
+                if (_logger.IsInfo) _logger.Info($"Deleting old block {number} with hash {hash}.");
+                _blockTree.DeleteBlock(number, hash, null, batch, null, true);
+                _receiptStorage.RemoveReceipts(block);
+                _deletePointer = number;
+                deletedBlocks++;
+            }
+        }
+        finally
+        {
+            SaveDeletePointer();
+            if (_logger.IsInfo) _logger.Info($"Completed pruning operation up to timestamp {cutoffTimestamp}. Deleted {deletedBlocks} blocks up to #{_deletePointer}.");
+        }
+    }
+
+    private IEnumerable<Block> GetBlocksBeforeTimestamp(ulong cutoffTimestamp)
+        => GetBlocksByNumber(_deletePointer, b => b.Timestamp >= cutoffTimestamp, i => _deletePointer = i);
+
+    private IEnumerable<Block> GetBlocksByNumber(long from, Predicate<Block> endSearch, Action<long> onFirstBlock)
+    {
+        bool firstBlock = true;
+        long headNumber = _blockTree.Head!.Number;
+        for (long i = from; i <= headNumber; i++)
+        {
+            ChainLevelInfo? chainLevelInfo = _chainLevelInfoRepository.LoadLevel(i);
+            if (chainLevelInfo is null)
+            {
+                continue;
+            }
+            else if (firstBlock)
+            {
+                onFirstBlock(i);
+                firstBlock = false;
+            }
+
+            bool finished = false;
+            foreach (BlockInfo blockInfo in chainLevelInfo.BlockInfos)
+            {
+                Block? block = _blockStore.Get(blockInfo.BlockNumber, blockInfo.BlockHash);
+                if (block is null)
+                {
+                    continue;
+                }
+
+                // search entire chain level before finishing
+                if (endSearch(block))
+                {
+                    finished = true;
+                }
+                else
+                {
+                    yield return block;
+                }
+            }
+
+            if (finished)
+            {
+                break;
+            }
+        }
+    }
+
+    private ulong CalculateCutoffTimestamp()
+    {
+        ulong cutoffTimestamp = 0;
+
+        if (_historyConfig.HistoryRetentionEpochs.HasValue)
+        {
+            cutoffTimestamp = _blockTree.Head!.Timestamp - (ulong)(_historyConfig.HistoryRetentionEpochs.Value * _epochLength);
+        }
+
+        if (_historyConfig.DropPreMerge)
```

Comment:
`DropPreMerge` has priority over `HistoryRetentionEpochs`? Or the other way round?